designs/2024-hooks-for-test-cases/README.md

```diff
@@ -0,0 +1,308 @@
+  jest.resetAllMocks(); // Explicitly clean up any mocks or state changes
+}
+```
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+
+The new `setup` property should be formally documented in the following ways:
+
+- ESLint Rule Tester Documentation: Update the official ESLint `RuleTester` documentation to include detailed
+  information about the new property. This documentation should explain:
+  - The purpose and use cases of the setup property;
+  - How to implement the setup function in test cases;
+  - Example demonstrating the usage of the `setup` property for both `valid` and `invalid` test cases;
+- Changelog: Include an entry in the ESLint changelog outlining the new property with a brief description;
+
+### Announcement
+
+To ensure that the ESLint community is aware of the new feature and understands its motivation and usage, a formal
+announcement should be made on the ESLint blog. This announcement can include:
+
+- Introduction to the Feature: Explain what the setup property is and provide context as to why it was introduced;
+- Motivation: Describe the motivation for introducing the setup property, including the challenges with the current
+  approach and the anticipated benefits of the new feature;
+- Examples and Use Cases: Provide concrete examples and use cases to show how the setup property can be used to
+  improve the testing process for ESLint plugins;
+- Link to Documentation: Include links to the updated documentation having more detailed information.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think 
+    about any opposing viewpoints that reviewers might bring up. 
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+Possible concerns:
+
+- Potential for Misuse:
+  - Despite being an optional feature, there remains a risk that some users might misuse the setup property
+    by inadvertently modifying global state, leading to flaky tests.
+- Performance:
+  - There is a risk that developers might overload the setup property, leading to performance decrease.
+- Redundancy: confusion with `beforeEach`:
+  - It is important to emphasize the key difference between these entities in the documentation:
+    - `beforeEach` is for **consistent** actions applicable to **all** tests;
+    - `setup` is for **specific** actions applicable to **particular** test.
+- Limited Use Case:
+  - The necessity for the `setup` might be limited to specific rules and plugins. The broader ESLint user
+    community may not frequently encounter these challenges, and it may seem unjustified to them.
```

Comment:
A note that `RuleTester` typically reports an error if it founds duplicate test cases. Test cases with a `setup` function would be [excluded from the check](https://github.com/eslint/eslint/blob/4b23ffd6454cfb1a269430f5fe28e7d1c37b9d3e/lib/rule-tester/rule-tester.js#L872-L873).

Reply:
But functions are serializable, @fasttime 

<img width="400" alt="Screenshot 2024-08-14 at 08 41 05" src="https://github.com/user-attachments/assets/e85815a0-53a6-4130-ac1b-14c3e0a305fc">

Reply:
though, not by `JSON.stringify`

Reply:
it's a good point, nice catch

Reply:
What if I include `setup` to the `duplicationIgnoredParameters` list, @fasttime ?
Should it retain the operation of detecting duplicates?

https://github.com/eslint/eslint/blob/9bde90c2edb6800c7f6428c5550ff00fff44ab02/lib/rule-tester/rule-tester.js#L143-L147

Reply:
Yeah, the algorithm we have in place to check if a test case is eligible for serialization does not like functions, see https://github.com/eslint/eslint/blob/2e1d54960051b59e1c731fa44c2ef843290b1335/lib/shared/serialization.js#L25-L51.

A discussion around this limitation can be found in https://github.com/eslint/rfcs/pull/84#discussion_r775150885.

Reply:
I'd suggest to switch the serialization to this library: https://www.npmjs.com/package/serialize-javascript
`isSerializable` check can be removed then.

Reply:
Are you thinking to change the way serialization works, or will `serialize-javascript` offer the same behavior, just without the self-written logic of `isSerializable`?

Reply:
As per my understanding, serialization is only needed to compare test cases in order to find duplicates, which seem to be a major feature that should be retained.
Since the current serialization method is `JSON.stringify` (which is designed to be JSON compatible, while JSON format itself is not utilized), I'm thinking on improvement of the serialization approach by involving that or similar library which is capable for serializing all JavaScript entities including function (unrelated to JSON format but suitable exactly for the purpose of comparison).
The proprietary `isSerializable` function then can be removed because the mentioned library does not have such limitations as JSON, @fasttime

Reply:
Regular expression and function are excluded from serialization for a purpose as noted in https://github.com/eslint/rfcs/pull/84#discussion_r775150885. It's also noted in [that RFC](https://github.com/eslint/rfcs/blob/9a86b5e216263f39ccd01911f2b38d53ce9d10ab/designs/2021-stricter-rule-test-validation/README.md) that

> Detecting duplicates isn't critical, so it's okay that we will skip some test cases, but we could potentially make this smarter later.

So I think it's acceptable if test cases with a `setup` function are not checked for duplicates. We could make the serialization logic smarter if we can address the concerns expressed in that RFC, but that should not a blocker for this change in my opinion.

Reply:
ah, ok then, thank you for the clarification.
I will reflect it in the RFC.